### PR TITLE
stdenv: improve build phases output

### PR DESCRIPTION
**Pull request by imincik**
## Description of changes

In the current state, it is not possible to know which build phases and in which order where used for the build process. This information is very useful when working with packages, which are using customized phases (for example Python packages built using `buildPythonPackage`).

Also, it is not possible to know which line of the build output belongs to which phase.

This change will:

1. print list of build phases at the beginning of the build process

For example:
```
phases: unpackPhase patchPhase updateAutotoolsGnuConfigScriptsPhase configurePhase buildPhase checkPhase installPhase fixupPhase installCheckPhase distPhase
```

2. improve consistency of phase start and finish messages, by always printing them in the same format for all phases.

Message format: `started phase: <PHASE-NAME>`or `finished phase: <PHASE-NAME>`

For example:
```
started unpackPhase

unpacking source archive /nix/store/447hvnlzzi9myri1iq3bijxgx6v6b592-patchelf-0.15.0.tar.bz2
source root is patchelf-0.15.0
setting SOURCE_DATE_EPOCH to timestamp 1657949009 of file patchelf-0.15.0/patchelf.spec

finished unpackPhase
```

These small changes will improve readability of build process log and will allow better understanding of the build process. Knowing the list of phases and their execution order is needed for package development using `nix develop .#<PACKAGE>`.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/stdenv/generic/setup.sh [hunk 1]
```diff
@@ -1508,17 +1508,7 @@ distPhase() {
...
-        patchPhase) echo "patching sources";;
-        configurePhase) echo "configuring";;
-        buildPhase) echo "building";;
-        checkPhase) echo "running tests";;
-        installPhase) echo "installing";;
-        fixupPhase) echo "post-installation fixup";;
-        installCheckPhase) echo "running install tests";;
-        *) echo "$phase";;
-    esac
+        echo "started $phase"
```
Comment (lorenzleutgeb) [inline pkgs/stdenv/generic/setup.sh:1511]: 1. I think you should explicitly print the word "phase". I am not sure which lines are printed before/after, but just "started X" might be less informative than "started phase X". Also I have tried many times to just search for "phase" in output to see what happens in which phase. By printing the word "phase", iterating through matches for "phase" would enable that.
 
```suggestion
        echo "started phase: $phase"
```

 2. I am not sure whether "started"/"finished" is the canonical terminology regarding phases. If it is, then okay, I won't complain. But I personally feel that "finished" should be used sparingly, and I also think that "begin"/"end" or "start"/"end" would be better.
 
```suggestion
        echo "begin phase: $phase"
```
```suggestion
        echo "beginning of phase: $phase"
```
```suggestion
        echo "start phase: $phase"
```
```suggestion
        echo "start of phase: $phase"
```

File: pkgs/stdenv/generic/setup.sh [hunk 2]
```diff
@@ -1527,6 +1517,7 @@ showPhaseFooter() {
     local startTime="$2"
     local endTime="$3"
     local delta=$(( endTime - startTime ))
+    echo "finished $phase"
```
Comment (lorenzleutgeb) [inline pkgs/stdenv/generic/setup.sh:1520]:
```suggestion
        echo "end phase: $phase"
```
```suggestion
        echo "end of phase: $phase"
```

**Review by lorenzleutgeb — CHANGES_REQUESTED** [review]

File: pkgs/stdenv/generic/setup.sh (hunk 1 above)
Comment (imincik) [reply pkgs/stdenv/generic/setup.sh:1511]: @lorenzleutgeb , thank you very much for review. I agree adding `phase` word to the message is a good idea.

**Comment by imincik** [comment]
I updated messages as suggested. Thanks

**Comment by bjornfor** [comment]
CC @maralorn due to https://github.com/maralorn/nix-output-monitor (awesome tool!), which seem able to detect the phases already. (Via JSON output from Nix?)

**Comment by imincik** [comment]
> CC @maralorn due to https://github.com/maralorn/nix-output-monitor (awesome tool!), which seem able to detect the phases already. (Via JSON output from Nix?)

Yes, I know and use nom. It is really great tool, but still this PR makes a sense.

**Comment by lorenzleutgeb** [comment]
> CC @maralorn due to https://github.com/maralorn/nix-output-monitor (awesome tool!), which seem able to detect the phases already. (Via JSON output from Nix?)

Yes, with `--output-format internal-json` you'll get lines like:

```
@nix {"action":"result","fields":["fixupPhase"],"id":4452790164258821,"type":104}
```

and the ID 104 is what `nom` ingests, see https://github.com/maralorn/nix-output-monitor/blob/6ca91b73f9e51b4486b3542c23d496f470026e2b/lib/NOM/Parser/JSON.hs#L115

I think this PR is intended to improve using `nix` without `nom`, and printing phases is helpful as such. So how is `nom` relevant?

**Review by lorenzleutgeb — APPROVED** [review]

**Comment by bjornfor** [comment]
> I think this PR is intended to improve using nix without nom, and printing phases is helpful as such. So how is nom relevant?

It seemed relevant to me. Sorry if it was noise.

File: pkgs/stdenv/generic/setup.sh [hunk 3]
```diff
@@ -1527,6 +1517,7 @@ showPhaseFooter() {
     local startTime="$2"
     local endTime="$3"
     local delta=$(( endTime - startTime ))
+    echo "finished phase: $phase"
```
Comment (Artturin) [inline pkgs/stdenv/generic/setup.sh:1520]: this is a dupe of the `completed in` no?

File: pkgs/stdenv/generic/setup.sh [hunk 4]
```diff
@@ -1508,17 +1508,7 @@ distPhase() {
...
-        unpackPhase) echo "unpacking sources";;
-        patchPhase) echo "patching sources";;
-        configurePhase) echo "configuring";;
-        buildPhase) echo "building";;
-        checkPhase) echo "running tests";;
-        installPhase) echo "installing";;
-        fixupPhase) echo "post-installation fixup";;
-        installCheckPhase) echo "running install tests";;
-        *) echo "$phase";;
-    esac
```
Comment (Artturin) [inline pkgs/stdenv/generic/setup.sh:1521]: the phase is echoed here so it's already possible to know which phase what commands are run in

File: pkgs/stdenv/generic/setup.sh (hunk 3 above)
Comment (Artturin) [reply pkgs/stdenv/generic/setup.sh:1520]: If we're adding a print like this always then it makes sense to add the completed in to it. The reason I limited it to >30 is because i wanted to avoid the extra line

Comment (Artturin) [inline pkgs/stdenv/generic/setup.sh:1520]:
```suggestion
    echo -n "finished phase: $phase in "
```

Comment (imincik) [reply pkgs/stdenv/generic/setup.sh:1520]: @Artturin , thank you for review. I am not sure if I understand your suggestion. Do you want to always print `$phase completed in` ? If that's true, then we need to remove line saying `(( delta < 30 )) && return`.

Comment (lorenzleutgeb) [reply pkgs/stdenv/generic/setup.sh:1520]: IMO it's worth printing the stage unconditionally.

File: pkgs/stdenv/generic/setup.sh [hunk 5]
```diff
@@ -1558,6 +1549,7 @@ genericBuild() {
             configurePhase ${preBuildPhases[*]:-} buildPhase checkPhase \
             ${preInstallPhases[*]:-} installPhase ${preFixupPhases[*]:-} fixupPhase installCheckPhase \
             ${preDistPhases[*]:-} distPhase ${postPhases[*]:-}";
+        echo "phases: $(echo $phases | tr -s ' ')"
```
Comment (lorenzleutgeb) [inline pkgs/stdenv/generic/setup.sh:1552]: To avoid a subshell, use:
```suggestion
        echo -n "phases: "
        echo $phases | tr -s ' '
```

**Review by lorenzleutgeb — CHANGES_REQUESTED** [review]

File: pkgs/stdenv/generic/setup.sh (hunk 3 above)
Comment (Artturin) [reply pkgs/stdenv/generic/setup.sh:1520]: either we print in it always on the same line or on the same line when the time is above 30 seconds like now

File: pkgs/stdenv/generic/setup.sh (hunk 5 above)
Comment (imincik) [reply pkgs/stdenv/generic/setup.sh:1552]: Actually, following works as well and looks better, but thanks for thinking about performance as well

```
echo "phases: $phases" | tr -s ' '
```

**Review by lorenzleutgeb — APPROVED** [review]

File: pkgs/stdenv/generic/setup.sh (hunk 3 above)
Comment (imincik) [reply pkgs/stdenv/generic/setup.sh]: Thanks,

I changed `showPhaseFooter` output to always duration of the phase. 

Now we have output like this:
```
nix-build -A hello 2>&1 | grep "finished phase"

finished phase: unpackPhase in 0 seconds
finished phase: patchPhase in 0 seconds
finished phase: updateAutotoolsGnuConfigScriptsPhase in 0 seconds
finished phase: configurePhase in 13 seconds
finished phase: configurePhase in 1 minutes 2 seconds
finished phase: buildPhase in 15 seconds
```

It looks to me that always having duration info can be be helpful for debugging. I hope it doesn't slow things down.

**Comment by maralorn** [comment]
The way nom gets the phase information is incredibly weird because some code, and I don't remember where exactly but I think in nixpkgs and not in nix, injects a fake internal json message. I forgot exactly how it works. But maybe someone could check whether this PR breaks phase reporting in nom?

**Comment by maralorn** [comment]
Ah, okay the setup.sh prints an `@nix { ... }` line right before showPhaseHeader is called. So that is completely unaffected.

Otherwise nice PR, thanks!

File: pkgs/stdenv/generic/setup.sh [hunk 6]
```diff
@@ -1508,31 +1508,20 @@ distPhase() {
...
     local endTime="$3"
-    local delta=$(( endTime - startTime ))
-    (( delta < 30 )) && return
 
+    local delta=$(( endTime - startTime ))
     local H=$((delta/3600))
     local M=$((delta%3600/60))
     local S=$((delta%60))
-    echo -n "$phase completed in "
+    echo -n "finished phase: $phase in "
```
Comment (Artturin) [inline pkgs/stdenv/generic/setup.sh:1524]: Honestly do we need this always because it's very easy to tell when a phase ended with just the header

```
gnu-config> phases: unpackPhase patchPhase configurePhase buildPhase checkPhase installPhase fixupPhase installCheckPhase distPhase
gnu-config> started phase: unpackPhase
gnu-config> finished phase: unpackPhase in 0 seconds
gnu-config> started phase: patchPhase
gnu-config> finished phase: patchPhase in 0 seconds
gnu-config> started phase: installPhase
gnu-config> finished phase: installPhase in 0 seconds
gnu-config> started phase: fixupPhase
```

```
bison> phases: unpackPhase patchPhase updateAutotoolsGnuConfigScriptsPhase configurePhase buildPhase checkPhase installPhase fixupPhase installCheckPhase distPhase
bison> started phase: unpackPhase
bison> unpacking source archive /nix/store/9vm1ihdg1ysmrjdbb80g834iizzxb4yk-bison-3.8.2.tar.gz
bison> source root is bison-3.8.2
bison> setting SOURCE_DATE_EPOCH to timestamp 1632561040 of file bison-3.8.2/src/parse-gram.output
bison> finished phase: unpackPhase in 0 seconds
bison> started phase: patchPhase
bison> finished phase: patchPhase in 0 seconds
bison> started phase: updateAutotoolsGnuConfigScriptsPhase
bison> Updating Autotools / GNU config script to a newer upstream version: ./build-aux/config.sub
bison> Updating Autotools / GNU config script to a newer upstream version: ./build-aux/config.guess
bison> finished phase: updateAutotoolsGnuConfigScriptsPhase in 0 seconds
bison> started phase: configurePhase
```

Kinda cluttery

Comment (lorenzleutgeb) [reply pkgs/stdenv/generic/setup.sh:1524]: If the phases are not doing anything, then this invites to set e.g. `dontUnpack = true;`, `dontPatch = true;`, `dontFixup = true;`...

Comment (Artturin) [reply pkgs/stdenv/generic/setup.sh:1524]: Often they're doing something but it's so fast and not logged that it's not visible

We have NIX_DEBUG=7 to see all the unimportant debug stuff

Some dontX should not be set without considerations because what if in the future something does something in that phase.

File: pkgs/stdenv/generic/setup.sh [hunk 7]
```diff
@@ -1558,6 +1547,7 @@ genericBuild() {
             configurePhase ${preBuildPhases[*]:-} buildPhase checkPhase \
             ${preInstallPhases[*]:-} installPhase ${preFixupPhases[*]:-} fixupPhase installCheckPhase \
             ${preDistPhases[*]:-} distPhase ${postPhases[*]:-}";
+        echo "phases: $phases" | tr -s ' '
```
Comment (Artturin) [inline pkgs/stdenv/generic/setup.sh:1550]: This also shows the phases which have been disabled with `dontX` because the `dontX` logic is done in the `for curPhase` loop

File: pkgs/stdenv/generic/setup.sh [hunk 8]
```diff
@@ -1508,31 +1508,20 @@ distPhase() {
...
-        patchPhase) echo "patching sources";;
-        configurePhase) echo "configuring";;
-        buildPhase) echo "building";;
-        checkPhase) echo "running tests";;
-        installPhase) echo "installing";;
-        fixupPhase) echo "post-installation fixup";;
-        installCheckPhase) echo "running install tests";;
-        *) echo "$phase";;
-    esac
+        echo "started phase: $phase"
```
Comment (Artturin) [inline pkgs/stdenv/generic/setup.sh:1511]: :+1: on removing the special casing of certain phases, would be fine to split this in to it's own PR and merging as.

I looked through the github search for a few of the strings and couldn't see anyone using them.

**Comment by imincik** [comment]
> Ah, okay the setup.sh prints an `@nix { ... }` line right before showPhaseHeader is called. So that is completely unaffected.
> 
> Otherwise nice PR, thanks!

Yes, I test this PR with nom as well. Thanks for the comment.

File: pkgs/stdenv/generic/setup.sh (hunk 6 above)
Comment (imincik) [reply pkgs/stdenv/generic/setup.sh:1524]: Footer message might looks redundant, but what I don't like is it footer message is not printed consistently - it is printed only if phase execution takes longer time. I think I would like either dropping footer message completely or printing it all the time to avoid any confusion for the log reader.

Comment (maralorn) [reply pkgs/stdenv/generic/setup.sh:1524]: I wonder if footer message and header of next phase could be merged into one line?

File: pkgs/stdenv/generic/setup.sh (hunk 7 above)
Comment (Artturin) [reply pkgs/stdenv/generic/setup.sh:1550]: @imincik this doesn't seem resolved?

File: pkgs/stdenv/generic/setup.sh (hunk 6 above)
Comment (imincik) [reply pkgs/stdenv/generic/setup.sh:1524]: @maralorn , thank you for contributing to this discussion.

I personally, don't like this idea because it would complicate implementation little bit and can make navigation in the output little bit harder for the log reader. I see only little advantage of this which is reduction of few lines in the build output which is usually containing few hundreds or thousands of lines.

File: pkgs/stdenv/generic/setup.sh (hunk 7 above)
Comment (imincik) [reply pkgs/stdenv/generic/setup.sh:1550]: Sorry, thanks for re-opening.

Comment (imincik) [reply pkgs/stdenv/generic/setup.sh]: @Artturin , what about [this change](https://github.com/NixOS/nixpkgs/pull/257229/commits/288756e6017c9e036e2f11863cff6a8c5d03680f).

* loop over the phases to get the list of those which are not disabled - `activePhases`
* print list of `activePhases`
* loop over the `activePhases` in order to execute them ?

Comment (imincik) [reply pkgs/stdenv/generic/setup.sh]: @Artturin , would you please review my changes ? Thank you very much.

**Comment by imincik** [comment]
Can I kindly as people interested in this PR to re-review this PR until it has a change to be merged in 23.11 ? Thank you all !

File: pkgs/stdenv/generic/setup.sh (hunk 6 above)
Comment (RaitoBezarius) [reply pkgs/stdenv/generic/setup.sh:1524]: Do we have consensus on this @Artturin @lorenzleutgeb @maralorn ?
Given that we have kinda 886G of build logs, I would be in favor of anything that contributes towards slowing down the rate of creating more bytes to the build logs.

**Comment by amjoseph-nixpkgs** [comment]
I don't mean to derail the discussion here, but:

One way to dodge the bikeshedding would be to add an `--option log-filter` to `nix`.  Whatever you set that option to is run in a separate sandbox; it receives the builder's stdout/stderr, and its stdout/stderr is what is written to the logfile.

The important point here is that you can change `--option log-filter` freely without affecting purity, since it has no way to influence the outpath resulting from the build.

Once that's in place, we can log as much as we want from `setup.sh`, and people who don't like the extra logging (or Hydra where excessive logs are a problem) can just filter it.  The bikeshedding discussions would be moved to the default logger, which can be changed **without a mass-rebuild**.

File: pkgs/stdenv/generic/setup.sh (hunk 6 above)
Comment (lorenzleutgeb) [reply pkgs/stdenv/generic/setup.sh:1524]: It seems that there's agreement on unconditionally printing the start of every phase, but that there is no agreement on printing the end of every phase (my preference is to unconditionally print it, but I personally don't have to deal with hundreds of gibibytes of build logs...). A compromise that I'd agree with would be to leave the logic for printing the end of a phase unchanged, merge this PR, and then continue a separate discussion towards https://github.com/NixOS/nixpkgs/pull/257229#issuecomment-1774259165.

**Comment by lorenzleutgeb** [comment]
> One way to dodge the bikeshedding would be to add an `--option log-filter` to `nix`. Whatever you set that option to is run in a separate sandbox; it receives the builder's stdout/stderr, and its stdout/stderr is what is written to the logfile.

Great idea! Maybe it's worth considering to not to use plain stdout/stderr as input, but the ["internal-json" log format](https://nixos.org/manual/nix/stable/command-ref/nix-build#opt-log-format) that tools like @maralorn's [nix-output-monitor](https://github.com/maralorn/nix-output-monitor) use succesfully.

**Comment by imincik** [comment]
> One way to dodge the bikeshedding would be to add an --option log-filter to nix.

I like this idea too. Is anybody keen to work on this feature ?

**Comment by imincik** [comment]
Extract of changes in this PR submitted in https://github.com/NixOS/nixpkgs/pull/267639
